smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptJmesPathVisitor.java

```diff
@@ -0,0 +1,315 @@
+
+    @Override
+    public Void visitMultiSelectList(MultiSelectListExpression expression) {
+        ArrayList<String> evaluators = new ArrayList<String>();
+
+        String executionContextInital = this.executionContext;
+
+        expression.getExpressions().forEach((JmespathExpression exp) -> {
+            exp.accept(this);
+            evaluators.add(this.executionContext);
+            this.executionContext = executionContextInital;
+        });
+
+        String resultScope = makeNewScope("result_");
+        writer.write("let $L = [];", resultScope);
+        for (String evaluator : evaluators) {
+            writer.write("$L.push($L);", resultScope, evaluator);
+        }
+        writer.write("$L = $L;", this.executionContext, resultScope);
+
+        return null;
+    }
+
+    @Override
+    public Void visitMultiSelectHash(MultiSelectHashExpression expression) {
+        throw new CodegenException("TypeScriptJmesPath visitor not implemented MultiSelectHashExpression");
+    }
+
+    @Override
+    public Void visitAnd(AndExpression expression) {
+        throw new CodegenException("TypeScriptJmesPath visitor not implemented AndExpression");
+    }
+
+    @Override
+    public Void visitOr(OrExpression expression) {
+        throw new CodegenException("TypeScriptJmesPath visitor not implemented OrExpression");
+    }
+
+    @Override
+    public Void visitNot(NotExpression expression) {
+        throw new CodegenException("TypeScriptJmesPath visitor not implemented NotExpression");
+    }
+
+    @Override
+    public Void visitObjectProjection(ObjectProjectionExpression expression) {
+        expression.getLeft().accept(this);
+        String element = makeNewScope("element_");
+        String result = makeNewScope("result_");
+
+        writer.openBlock("let $L = Object.values($L).map($L => {", "});", result,
+                this.executionContext, element, () -> {
+            this.executionContext = element;
+            expression.getRight().accept(this);
+            writer.write("return $L;", this.executionContext);
+        });
+        this.executionContext = result;
+        return null;
+    }
+
+    @Override
+    public Void visitProjection(ProjectionExpression expression) {
+        expression.getLeft().accept(this);
+        String element = makeNewScope("element_");
+        String result = makeNewScope("result_");
+
+        writer.openBlock("let $L = $L.map($L => {", "});", result,
+                this.executionContext, element, () -> {
+            this.executionContext = element;
+            expression.getRight().accept(this);
+            writer.write("return $L;", this.executionContext);
+        });
+        this.executionContext = result;
+        return null;
+    }
+
+    @Override
+    public Void visitFilterProjection(FilterProjectionExpression expression) {
+        expression.getLeft().accept(this);
+
+        expression.getRight().accept(this);
+
+        String filterScope = makeNewScope("scope_");
```

Comment:
"filter" would be far more informative in the generated code than "scope".

Reply:
done